tree/dataframe/inc/ROOT/RSqliteDS.hxx

```diff
@@ -0,0 +1,87 @@
+// Author: Jakob Blomer CERN  07/2018
+
+/*************************************************************************
+ * Copyright (C) 1995-2017, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RSQLITEDS
+#define ROOT_RSQLITEDS
+
+#include "ROOT/RDataFrame.hxx"
+#include "ROOT/RDataSource.hxx"
+
+#include <map>
+#include <memory>
+#include <mutex>
+
+#include <sqlite3.h>
+
+namespace ROOT {
+
+namespace RDF {
+
+class RSqliteDS final : public ROOT::RDF::RDataSource {
+private:
+   enum class Types {
+     kInteger,
+     kReal,
+     kText,
+     kBlob,
+     kNull
+   };
+
+   struct Value_t {
+     Value_t() : fType(Types::kNull), fNull(nullptr), fPtr(nullptr), fIsActive(false) {}
+
+     Types fType;
+     Long64_t fInteger;
+     double fReal;
+     std::string fText;
+     std::vector<unsigned char> fBlob;
+     void *fNull;
```

Comment:
It's actually the standard SQL `NULL`, meaning "I don't know". In SQL operations, a single `NULL` operand always taints the entire result to be `NULL` except for the boolean operation `IS NULL`.  Mapping this sort of three-valued logic to C++ is a bit tricky, I think.  I thought about adding a wrapper class for each the sqlite types (`int`, `float`, `string`, etc.) that could accept a value or `NULL` and to overload all the operators for these wrapper classes but it seemed overkill.  In its current implementation, the data source tries to figure out the real type of every column, and if a `NULL` value occurs, it is converted to the neutral element of that type (`0`, `0.0`, `""`, etc).  Only if it can't figure out the type of a column (4th point in the PR comment), the `Types::kNull` case is ever used.

Users can't do anything meaningful with a value of type `Types::kNull`.  It is most likely an indication of an ill-queried data set.

Setting `fPtr` directly to `nullptr` creates problems with `GetColumnReadersImpl()`, where the address of `fPtr` is returned.

Reply:
I see, thank you for the explanation!